srcs/main.c

```diff
@@ -6,17 +6,324 @@
-/*   Updated: 2017/05/11 16:56:14 by fgallois         ###   ########.fr       */
+/*   Updated: 2017/06/09 11:49:49 by fgallois         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
-#include <unistd.h>
-#include <SDL.h>
+# include "../include/rtv1.h"
+#   include "../include/event.h"
+# include <SDL2/SDL.h>
+# include <stdlib.h>
+#include <stdio.h>
+# include <math.h>
 
-int		main(void)
+
+void		rotate_vector(t_env *env, t_vect *ray_dir)
+{
+	float plus_phi;
+	float plus_theta;
+
+	plus_phi = env->cam->add_phi;
+	plus_theta = env->cam->add_theta;
+	change_vect(ray_dir, plus_phi, plus_theta);
+}
+
+t_ray		*current_ray(t_env *env, t_vect *v)
+{
+	t_ray *r;
+	t_vect *mini;
+	t_vect *w;
+	t_vect *ray_dir;
+	t_vect  *new_pos_cam;
+
+	new_pos_cam = add_vect(env->cam->pos, env->cam->trans);
+	mini = minus_vect(v, new_pos_cam);
+	w = new_vect(0.0, 0.0, 0.0);
+	ray_dir = mini;
+
+	rotate_vector(env, ray_dir);
+    normed(ray_dir);
+	r = new_ray(new_pos_cam, ray_dir, norm(mini), w);
+	return (r);
+}
+
+int            find_nearest_inter(t_env *env, t_vect *v, t_hit_point **mem, t_obj **colore)
 {
-	SDL_Init(SDL_INIT_VIDEO);
+	t_ray *r;
+	t_obj *tmp;
+	t_hit_point *hp;
+	float min;
 
-	SDL_Quit();
+	tmp = env->obj;
+	min = INFINI;
+	hp = NULL;
+	r = current_ray(env, v);
+	while (tmp)
+	{   
+		if ((hp = tmp->is_hit(tmp->type, r)))
+		{
+			hp->distance_to_cam = distance_with_cam(env, hp);
+			if (hp->distance_to_cam < min)
+			{
+				min = hp->distance_to_cam;
+				*colore = tmp;
+				*mem = hp;
+			}
+			else
+            	free(hp);
+		}
+		tmp = tmp->next;
+	}
+	free(r->origin);
+	free(r->color);
+	free(r);
+	return (min);
+}
+
+int        is_light_reached(t_light *light, t_env *env, t_hit_point *mem, t_obj *colore)
+{
+	t_vect *mini;
+	t_vect *w;
+	t_ray *r;
+	t_obj *tmp;
+	t_hit_point *hr;
+	int meet_object;
+
+	meet_object = 0;
+	mini = minus_vect(light->pos, mem->vect);
+	normed(mini);
+	w = new_vect(0.0, 0.0, 0.0);
+	r = new_ray(mem->vect, mini, 0.0, w);
+	tmp = env->obj;
+	while (tmp && !meet_object)
+	{   
+		if (tmp == colore)
+			tmp = tmp->next;
+		else if (!(hr = tmp->is_hit(tmp->type, r)))
+			tmp = tmp->next;
+		else
+		{
+			if (distance(mem->vect, hr->vect) < distance(hr->vect, light->pos))
+				return (1);
+			tmp = tmp->next;
+			free(hr->vect);
+			free(hr->normal);
+			free(hr);
+		}
+	}
+    free(w);
+	free(r);
+	free(mini);
+	free(tmp);
 	return (0);
 }
+
+void               put_on_light(t_env *env, t_hit_point *mem, t_obj *colore, int p, int q)
+{
+    t_light *light;
+	t_vect *col;
+	float nb_of_lights;
+	int meet_object;
+
+	light = env->light;
+	nb_of_lights = numberoflights(env);
+	col = new_vect(0.0, 0.0, 0.0);
+	while (light)
+	{   
+		meet_object = is_light_reached(light, env, mem, colore);
+		if (!meet_object)
+			col = find_color_light(light, mem, colore->mater, col);
```

Comment:
Tu peux free en dehors de la fonction j'ai tords [](https://stackoverflow.com/questions/6054271/freeing-pointers-from-inside-other-functions-in-c)
Par contre tu ne peux pas malloc en dehors de la fonction
[](https://stackoverflow.com/questions/2513218/memory-allocated-with-malloc-does-not-persist-outside-function-scope)